File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.StringTokenizer;

public class Main {
    
    static int n;
    
    static int[][] arr;
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        n = Integer.parseInt(br.readLine());
        arr = new int[n][n];
        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int j = 0; j < i + 1; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        
        int max = arr[0][0];
        
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i + 1; j++) {
                arr[i][j] += Math.max(getRangeValue(i - 1, j), getRangeValue(i - 1, j - 1));
                max = Math.max(arr[i][j], max);
            }
        }
        
        System.out.println(max);
    }
    
    private static int getRangeValue(int x, int y) {
        if (x < 0 || x >= n || y < 0 || y >= n) {
            return 0;
        }
        return arr[x][y];
    }
    
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.js
function solution(triangle) {
    
    // 위에서 밑으로 더하는 방식으로 풀이 시 시간 초과 발생
    // 밑에서 위로 더하는 방식
    // 직각 삼각형 방향으로 봤을 때 더할 수의 방향은?
    // 바로 하위 요소 (x + 1, y) 또는 대각선 오른쪽 방향 요소 (x + 1, y + 1)
    const dirs = [[1, 0], [1, 1]]
    const mr = triangle.length
    
    for (let i = mr - 2; i >= 0; i--) {
        for (let j = 0; j < triangle[i].length; j++) {                          
            triangle[i][j] += Math.max(triangle[i + dirs[0][0]][j + dirs[0][1]], triangle[i + dirs[1][0]][j + dirs[1][1]])
        }
    }
    
    return triangle[0][0];
}